docs/architecture/adr-025-ibc-passive-channels.md

```diff
@@ -0,0 +1,126 @@
+
+In order to accomplish this, we propose adding more comprehensive events to expose channel metadata for each transaction sent from the `x/ibc/04-channel/keeper/handshake.go` and `x/ibc/04-channel/keeper/packet.go` modules.
+
+Here is an example of what would be in `ChanOpenInit`:
+
+```go
+const (
+  EventTypeChannelMeta = "channel_meta"
+  AttributeKeyAction = "action"
+  AttributeKeyHops = "hops"
+  AttributeKeyOrder = "order"
+  AttributeKeySrcPort = "src_port"
+  AttributeKeySrcChannel = "src_channel"
+  AttributeKeySrcVersion = "src_version"
+  AttributeKeyDstPort = "dst_port"
+  AttributeKeyDstChannel = "dst_channel"
+  AttributeKeyDstVersion = "dst_version"
+)
+// ...
+  // Emit Event with Channel metadata for the relayer to pick up and
+  // relay to the other chain
+  // This appears immediately before the successful return statement.
+  ctx.EventManager().EmitEvents(sdk.Events{
+    sdk.NewEvent(
+      types.EventTypeChannelMeta,
+      sdk.NewAttribute(types.AttributeKeyAction, "open_init"),
+      sdk.NewAttribute(types.AttributeKeySrcConnection, connectionHops[0]),
+      sdk.NewAttribute(types.AttributeKeyHops, strings.Join(connectionHops, ",")),
+      sdk.NewAttribute(types.AttributeKeyOrder, order.String()),
+      sdk.NewAttribute(types.AttributeKeySrcPort, portID),
+      sdk.NewAttribute(types.AttributeKeySrcChannel, chanenlID),
+      sdk.NewAttribute(types.AttributeKeySrcVersion, version),
+      sdk.NewAttribute(types.AttributeKeyDstPort, counterparty.GetPortID()),
+      sdk.NewAttribute(types.AttributeKeyDstChannel, counterparty.GetChannelID()),
+      // The destination version is not yet known, but a value is necessary to pad
+      // the event attribute offsets
+      sdk.NewAttribute(types.AttributeKeyDstVersion, ""),
+    ),
+  })
+```
+
+These metadata events capture all the "header" information needed to route IBC channel handshake transactions without requiring the client to query any data except that of the connection ID that it is willing to relay.  It is intended that `channel_meta.src_connection` is the only event key that needs to be indexed for a passive relayer to function.
+
+### Handling Channel Open Attempts
+
+In the case of the passive relayer, when one chain sends a `ChanOpenInit`, the relayer should inform the other chain of this open attempt and allow that chain to decide how (and if) it continues the handshake.  Once both chains have actively approved the channel opening, then the rest of the handshake can happen as it does with the current "naive" relayer.
+
+To implement this behavior, we propose adding a new callback `cbs.OnAttemptChanOpenTry` which explicitly handles the `MsgChannelOpenTry`, usually by resulting in a call to `keeper.ChanOpenTry`.  If the callback is not supplied, then the default behaviour would be to use `channel.HandleMsgChannelOpenTry`, for compatibility with existing chains that expect a "naive" relayer.
```

Comment:
Is there any other behavior from the callback other than calling the ChanOpenTry function and emitting the events from above? Maybe defining the function would help clarifying this

Reply:
I don't know how to write simple Go code to express what I have in mind, but essentially: 
`OnAttemptChanOpenTry` could return success without calling `ChanOpenTry`, and have the `ChanOpenTry` happen in a later block.  Think: I want to accept this channel, but only if another party pays for it.

Or: I need to wait on some external event to happen (possibly in another block) before I will know what version string I want to put into the `ChanOpenTry` request.

Reply:
That's fine, but then why not just wait until this other state change (e.g. payment) has happened before relaying the message at all? Then you can avoid the asynchronous processing. Is the idea that the relayer wouldn't want to track these sorts of state details?

Anyways, I think a callback which defaults to `keeper.ChanOpenTry` but which can do something else is fine.

Reply:
> Is the idea that the relayer wouldn't want to track these sorts of state details?

I'm mostly speaking of a model in which the relayer is generic and not dependent on these details, which can evolve separately on each chain (independent of the relayer).